src/dust.ts

```diff
@@ -0,0 +1,196 @@
+import { DustStateTime, dustStateTime } from "./state-time";
+
+/**
+ * Dust models
+ */
+export class Dust {
+    private readonly _Model: DustModelConstructable;
+    private readonly _particles: Particle[];
+    private readonly _random: Random;
+
+    /**
+     * @param Model A constructor for the model to use in the simulation
+     *
+     * @param pars Initial parameters to use with the model
+     *
+     * @param nParticles Number of particles to use in the simulation
+     * - can't be changed after construction.
+     *
+     * @param step Initial step
+     *
+     * @param random The random state
+     */
+    constructor(Model: DustModelConstructable, pars: Pars, nParticles: number,
+                step: number, random: Random) {
+        if (nParticles <= 0) {
+            throw Error("Expected at least one particle");
+        }
+        this._Model = Model;
+        this._particles = [];
+        this._random = random;
+
+        // NOTE: Here, we construct the model just once per dust
+        // object and then use exactly the same model object
+        // (including temporary internal storage) across all particles
+        // in object. This is in contrast with the approach taken in
+        // the C++ code where we kept separate copies so that they
+        // could be run in parallel. If we parallelise this with web
+        // workers we might need to revisit this approach, but the API
+        // there looks nothing like OpenMP and this would probably
+        // still be ok, once per worker.
+        const model = new Model(base, pars);
+        for (let i = 0; i < nParticles; ++i) {
+            this._particles.push(new Particle(model, step));
+        }
+    }
+
+    /**
+     * Returns the number of state elements per particle
+     */
+    public nState(): number {
+        return this._particles[0].size;
+    }
+
+    /**
+     * Returns the number of particles
+     */
+    public nParticles(): number {
+        return this._particles.length;
+    }
+
+    /**
+     * The current step
+     */
+    public step(): number {
+        return this._particles[0].step;
+    }
+
+    /**
+     * Returns information about how the state is packed. This is
+     * useful for computing indexes and extracting state.
+     */
+    public info(): DustModelInfo {
+        return this._particles[0].info();
+    }
+
+    /**
+     * Set new parameters into the model
+     *
+     * @param pars: New parameters
+     *
+     * @param setInitialState If `true`, then changing the parameters
+     * also updates the initial conditions to use those from the
+     * model's `initial()` method. This uses the current step of the
+     * model, so you may want to use {@link Dust.setStep} first.
+     */
+    public setPars(pars: Pars, setInitialState: boolean): void {
+        const step = this.step();
+        const nState = this.nState();
+        const model = new this._Model(base, pars);
+        if (model.size() !== this.nState()) {
```

Comment:
The size is the number of variables? so this error would occur if the new set of parameter cause the model to initialise with a different number of variables for some reason?

Reply:
yeah, this is a restriction more inherited from the C++ version where it simplified the bookkeeping and bought us some guarantees that help performance. For example, you can change parameter set mid-run without reallocating any of the storage you're using to accumulate results